compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -330,6 +359,62 @@ fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::Gen
     }
 }
 
+fn bidirectional_lifetime_predicates<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    item_def_id: LocalDefId,
+    bidir_outlives: impl Iterator<Item = (Lifetime, (LocalDefId, Symbol, Span))>,
+) -> Vec<(ty::Clause<'tcx>, Span)> {
+    let mut predicates = Vec::new();
+    let generics = tcx.generics_of(item_def_id.to_def_id());
+
+    bidirectional_lifetime_predicates_inner(
+        tcx,
+        item_def_id,
+        bidir_outlives,
+        generics,
+        &mut predicates,
+    );
+
+    predicates
+}
+
+fn bidirectional_lifetime_predicates_inner<'tcx>(
```

Comment:
This needs a comment that explains something like opaques have duplicated lifetimes and that we need these outlives predicates to enforce that these lifetimes stay in sync. Or maybe just uplift the comment that's on 327 right now.

```suggestion
fn compute_bidirectional_outlives_predicates<'tcx>(
```